arci-ros2/src/utils.rs

```diff
@@ -37,18 +24,45 @@ pub(crate) async fn wait(is_done: Arc<AtomicBool>) {
     }
 }
 
-pub(crate) async fn subscribe_one<T: Send + 'static>(
+pub(crate) fn subscribe_thread<T: Send + 'static, U: Send + Sync + 'static>(
     mut subscriber: impl Stream<Item = T> + Send + Unpin + 'static,
-) -> Result<Option<T>, tokio::task::JoinError> {
-    let is_done = Arc::new(AtomicBool::new(false));
-    let is_done_clone = is_done.clone();
-    let handle = tokio::spawn(async move {
-        let next = subscriber.next().await;
-        is_done.store(true, Ordering::Relaxed);
-        next
+    buf: Arc<RwLock<U>>,
+    mut f: impl FnMut(T) -> U + Send + 'static,
+) {
+    tokio::spawn(async move {
+        while Arc::strong_count(&buf) > 1 {
+            if let Some(val) = subscriber.next().await {
+                let res = f(val);
+                *buf.write().unwrap() = res;
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
     });
-    wait(is_done_clone).await;
-    handle.await
+}
+
+pub(crate) fn subscribe_one<T: Send>(
+    mut subscriber: impl Stream<Item = T> + Send + Unpin,
+) -> Option<T> {
+    // https://github.com/openrr/openrr/pull/501#discussion_r746183161
+    std::thread::scope(|s| {
+        s.spawn(|| {
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .unwrap()
+                .block_on(async move {
+                    for _ in 0..10 {
+                        if let Some(v) = subscriber.next().now_or_never() {
+                            return v;
+                        }
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                    }
+                    None // timeout
```

Comment:
I'm not sure how many seconds is a reasonable timeout, but I set it to 1s (100ms*10) for now.